File: src/Blocks/IO/PrintBlock.php
<?php

namespace Rhoban\Blocks\Blocks\IO;

use Rhoban\Blocks\Block;

abstract class PrintBlock extends Block
{
    /**
     * @see inherit
     */
    protected static $META = array(
        'name' => 'Print',
        'family' => 'I/O',
        'description' => 'Print all the inputs',
        'parameters' => array(
            array(
                'name' => 'Format',
                'type' => 'text',
                'card' => 0
            ),
            array(
                'name' => 'Inputs',
                'type' => 'integer',
                'default' => 1,
                'card' => 0
            )
        ),
        'inputs' => array(
            array(
                'name' => 'Trigger',
                'type' => 'integer',
                'card' => '0-1',
                'default' => 0
            ),
            array(
                'name' => 'Value #',
                'card' => '0-1',
                'length' => 'Inputs.value'
            ),
        ),
        'outputs' => array(),
    );
}

File: src/Interpreter/Cpp/templates/implementation/PrintBlock.cpp
void PrintBlock::initialize(Block *old)
{
}

void PrintBlock::tick()
{
    if (trigger) {
        int pos = 0;
        for (int i=0; i<format.size(); i++) {
            if (format[i] == '%' && i+1<format.size()) {
                if (format[i+1] == 'd') {
                    cout << (int)value[pos++];
                    i++;
                } else if (format[i+1] == 'f') {
                    cout << (double)value[pos++];
                    i++;
                } else {
                    cout << format[i];
                }
            } else {
                cout << format[i];
            }
        }
        cout << endl;
    }
}
